### PR TITLE
Add encoder/decoder and marshaler/unmarshaler for OTLP protobuf

### PR DESCRIPTION
**Pull request by bogdandrutu**
Updates https://github.com/open-telemetry/opentelemetry-collector/issues/3104

Signed-off-by: Bogdan Drutu <bogdandrutu@gmail.com>

### CODE REVIEW

File: internal/otlp/marshaler.go [hunk 1]
```diff
@@ -18,17 +18,32 @@ import (
...
 }
 
-// NewJSONLogsMarshaler returns a model.LogsMarshaler to decode from OTLP json bytes.
+// NewJSONLogsMarshaler returns a model.LogsMarshaler. Marshals to OTLP json bytes.
 func NewJSONLogsMarshaler() model.LogsMarshaler {
 	return model.NewLogsMarshaler(newJSONEncoder(), newFromTranslator())
 }
+
+// NewProtobufTracesMarshaler returns a model.TracesMarshaler. Marshals to OTLP binary protobuf bytes.
+func NewProtobufTracesMarshaler() model.TracesMarshaler {
```
Comment (bogdandrutu) [inline internal/otlp/marshaler.go:37]: @jrcamp should this be `Protobuf` or `Pb` or `Proto`?

**Review by carlosalberto — APPROVED** [review]

File: internal/otlp/marshaler.go (hunk 1 above)
Comment (jrcamp) [reply internal/otlp/marshaler.go:37]: I'd say `Protobuf`

File: internal/otlp/unmarshaler.go [hunk 2]
```diff
@@ -18,17 +18,32 @@ import (
 	"go.opentelemetry.io/collector/internal/model"
 )
 
-// NewJSONTracesUnmarshaler returns a model.TracesUnmarshaler to decode from OTLP json bytes.
+// NewJSONTracesUnmarshaler returns a model.TracesUnmarshaler. Unmarshalls from OTLP json bytes.
```
Comment (jrcamp) [inline internal/otlp/unmarshaler.go:21]:
```suggestion
// NewJSONTracesUnmarshaler returns a model.TracesUnmarshaler. Unmarshals from OTLP json bytes.
```

I think it's just 1 l (see https://golang.org/pkg/encoding/json/) and Google returns more results for Unmarshals.

File: internal/otlp/unmarshaler.go [hunk 3]
```diff
@@ -18,17 +18,32 @@ import (
...
 func NewJSONTracesUnmarshaler() model.TracesUnmarshaler {
 	return model.NewTracesUnmarshaler(newJSONDecoder(), newToTranslator())
 }
 
-// NewJSONMetricsUnmarshaler returns a model.MetricsUnmarshaler to decode from OTLP json bytes.
+// NewJSONMetricsUnmarshaler returns a model.MetricsUnmarshaler. Unmarshalls from OTLP json bytes.
```
Comment (jrcamp) [inline internal/otlp/unmarshaler.go:26]:
```suggestion
// NewJSONMetricsUnmarshaler returns a model.MetricsUnmarshaler. Unmarshals from OTLP json bytes.
```

File: internal/otlp/unmarshaler.go [hunk 4]
```diff
@@ -18,17 +18,32 @@ import (
...
 func NewJSONMetricsUnmarshaler() model.MetricsUnmarshaler {
 	return model.NewMetricsUnmarshaler(newJSONDecoder(), newToTranslator())
 }
 
-// NewJSONLogsUnmarshaler returns a model.LogsUnmarshaler to decode from OTLP json bytes.
+// NewJSONLogsUnmarshaler returns a model.LogsUnmarshaler. Unmarshalls from OTLP json bytes.
```
Comment (jrcamp) [inline internal/otlp/unmarshaler.go:31]:
```suggestion
// NewJSONLogsUnmarshaler returns a model.LogsUnmarshaler. Unmarshals from OTLP json bytes.
```

File: internal/otlp/pb_decoder.go [hunk 5]
```diff
@@ -0,0 +1,45 @@
...
+
+package otlp
+
+import (
+	otlpcollectorlogs "go.opentelemetry.io/collector/internal/data/protogen/collector/logs/v1"
+	otlpcollectormetrics "go.opentelemetry.io/collector/internal/data/protogen/collector/metrics/v1"
+	otlpcollectortrace "go.opentelemetry.io/collector/internal/data/protogen/collector/trace/v1"
+)
+
+type pbDecoder struct{}
```
Comment (jrcamp) [inline internal/otlp/pb_decoder.go:23]: Personally I'd just use protobuf instead of pb throughout.

**Review by jrcamp — APPROVED** [review]

File: internal/otlp/pb_decoder.go (hunk 5 above)
Comment (bogdandrutu) [reply internal/otlp/pb_decoder.go:23]: this is internal, so I don't think it is a problem :)

File: internal/otlp/marshaler.go (hunk 1 above)
Comment (bogdandrutu) [reply internal/otlp/marshaler.go:37]: Ok

File: internal/otlp/unmarshaler.go (hunk 2 above)
Comment (bogdandrutu) [reply internal/otlp/unmarshaler.go]: Done.

File: internal/otlp/unmarshaler.go (hunk 3 above)
Comment (bogdandrutu) [reply internal/otlp/unmarshaler.go]: Done.

File: internal/otlp/unmarshaler.go (hunk 4 above)
Comment (bogdandrutu) [reply internal/otlp/unmarshaler.go]: Done.

**Pull request merged**